File: android/src/main/kotlin/com/jarvanmo/nautilus/NautilusPlugin.kt
package com.jarvanmo.nautilus

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

class NautilusPlugin(registrar: Registrar) : MethodCallHandler {

    private val nautilusLoginService = LoginServiceHandler(registrar)
    private val tradeHandler = TradeHandler(registry = registrar)

    companion object {
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            val channel = MethodChannel(registrar.messenger(), "com.jarvanmo/nautilus")
            channel.setMethodCallHandler(NautilusPlugin(registrar))
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when {
            "login" == call.method -> nautilusLoginService.handleLogin(call, result)
            "initLogin" == call.method -> nautilusLoginService.initLogin(result)
            "isLogin" == call.method -> nautilusLoginService.handleIsLogin(result)
            "getUser" == call.method -> nautilusLoginService.handleGetUser(result)
            "logout" == call.method -> nautilusLoginService.handleLogout(result)
            "initTradeAsync" == call.method -> tradeHandler.initTradeAsync(call,result)
            "openItemDetail" == call.method -> tradeHandler.openItemDetail(call,result)
            "openUrl" == call.method -> tradeHandler.openUrl(call, result)
            "destroy" == call.method -> tradeHandler.destroy()
            else -> result.notImplemented()
        }
    }


}

File: ios/Classes/NautilusPlugin.m
#import "NautilusPlugin.h"


@implementation NautilusPlugin

NautilusLoginService *_loginService;
NautilusTradeHandler *_tradeHandler;


+ (void)registerWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel = [FlutterMethodChannel
            methodChannelWithName:@"com.jarvanmo/nautilus"
                  binaryMessenger:[registrar messenger]];
    NautilusPlugin *instance = [[NautilusPlugin alloc] initWithRegistrar:registrar];
    [registrar addMethodCallDelegate:instance channel:channel];
    [registrar addApplicationDelegate:instance];
}

- (instancetype)initWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    self = [super init];

    if (self) {
        _loginService = [[NautilusLoginService alloc] init];
        _tradeHandler = [[NautilusTradeHandler alloc] init];
    }

    return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([@"initTradeAsync" isEqualToString:call.method]) {
        [_tradeHandler initTradeAsync:call result:result];
    } else if ([@"login" isEqualToString:call.method]) {
        [_loginService handleLogin:call result:result];
    } else if ([@"isLogin" isEqualToString:call.method]) {
        [_loginService handleIsLogin:call result:result];
    } else if ([@"getUser" isEqualToString:call.method]) {
        [_loginService handleGetUser:call result:result];
    } else if ([@"logout" isEqualToString:call.method]) {
        [_loginService handleLogout:result];
    } else if([@"openItemDetail" isEqualToString:call.method]){
        [_tradeHandler openItemDetail:call result:result];
    } else if([@"openUrl" isEqualToString:call.method]){
        [_tradeHandler openUrl:call result:result];
    }else{
        result(FlutterMethodNotImplemented);
    }
}


- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {
    
    return [[AlibcTradeSDK sharedInstance] application:application openURL:url sourceApplication:sourceApplication annotation:annotation];
}

// NOTE: 9.0以后使用新API接口
- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<NSString *, id> *)options {
    return [[AlibcTradeSDK sharedInstance] application:app openURL:url options:options];
}

@end
